DIPs/DIP1xxx-RC.md

```diff
@@ -0,0 +1,232 @@
+
+### Reference
+
+There have been many conversations on D's NewsGroup attempting to suggest named arguments. For example [1](https://forum.dlang.org/post/khcalesvxwdaqnzaqotb@forum.dlang.org) and [2](https://forum.dlang.org/post/n8024o$dlj$1@digitalmars.com).
+
+Multiple library solutions have been attempted [1](https://forum.dlang.org/post/awjuoemsnmxbfgzhgkgx@forum.dlang.org) and [2](https://github.com/CyberShadow/ae/blob/master/utils/meta/args.d). Both of which do work but does not make the distinction between internal and public API arguments; which it cannot do without a lot more cruft.
+
+A [DIP](https://wiki.dlang.org/DIP88) (88) has been drafted, but never PR'd. Further work was done by Neia Neutuladh, but it has not been made public. At the time of writing Yuxuan Shui has a draft [DIP](https://github.com/yshui/DIPs/blob/master/DIPs/DIP1xxx-YS.md) in the review queue that is much more limited.
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Rationale
+
+Named arguments are a fairly popular language feature from dynamic languages which has been very highly requested on D's NewsGroup. It is available in Objective-C so it is also compatibility issue not just an enhancement for D.
+
+Unlike other languages however, this DIP does not aid in having all or even most arguments to be in _any_ order. The majority of arguments should remain ordered; this aids in readability and tooling. This removes a lot of potential problems which has been discussed and highly disliked within the D community about this concept.
+
+As a D only feature, it gives us more ways to describe interfaces in their usage but not what happens inside them. The decision to focus upon publically accessible aspects of an API using named arguments instead of their internals is to help make code clearer with preference of decreasing the number of template initations. A great example of this is ``max(a : 1, b : 3)`` is not more readable or understandable than ``max(1, 3)``. But ``MyInputRange!(Foo).Type`` is better than ``ElementType!(MyInputRange!(Foo))`` and describes that much more clearer that the type of the input range is clearly owned by ``MyInputRange`` not the mythical ``ElementType`` template.
+
+## Description
+
+This DIP adds a second kind of argument/parameter, named. Alternative names is ordered and unordered arguments. Ordered arguments is the current arrangement requiring that all argument be passed in the same order of the parameters defining them. But with unordered arguments, they can appear in any order or if specified to be optional at the parameter side, not at all.
+
+Named arguments are not affected by the passing of unnamed arguments. It does not matter where they go. So ``func(1, 2, o:true)`` is the same as ``func(1, o:true, 2)`` or ``func(o:true, 1, 2)``.

```

Comment:
How is ABI mangling for named arguments implemented? I presume it's by the argument namelist hash value, but I think it needs mentioning.

Reply:
No changes to mangling as it isn't required. Same goes for argument lookup, it isn't considered.

Reply:
Except when you're dealing with ``extern(C)``.

Name mangling is meant for one job only, keeping symbols unique in the binary image. I'm not against name mangling being affected, I just don't think its going to be an issue at this stage.

Reply:
It is an issue if it's not affected, because if you accidently write the named arguments differently in the header compared to the library, calling it will break the type system. (Except if we have a runtime check against that, but that would have a runtime cost badly fit for a systems programming language)

What comes to foreaign languge interfaces, I think we're better off not having named arguments there at all. Too dangerous.

Reply:
I didn't state it in the DIP but I did write: ``As a D only feature`` under rationale. It should only be valid with ``extern(D)``. But this does come under corner cases. Until somebody goes and start writing an implementation, I'm not sure that I can be convinced because the only way to experience the issues (as far as I'm aware)  that you're mentioning require some sort of sloppy coding. So I want to know that it isn't just something theoretical.

At any rate, the point I was making with ``extern(C)`` is that it'll be very easy in any systems programming language or comparable thereof to experience the same problems as you are currently presenting with or without named parameters.

Reply:
Yes, `extern (C)` is very bad in that regard, so let's not speard it's issues to `extern (D)`. And the risk of triggering this is very real. It takes only the following:

Both the library and a manually written interace correctly have a function:

export double foo(double a, double b)

The library author decides to extend the function to:

export double foo(double a, double b, <double[] tolerances = [0.05, 0.02, 0.01]>)

It takes just a library update and a slight overlook when updating the header to intoruce a bug in client code. Worse, since array is a reference type it will be a memory safety bug that can take ages to find.

Reply:
Yes but those 'header' files are meant to be generated by the compiler and not manually created.

Finally there is one example neither of us has so far brought up which will create the same stack corruption issues:

.di:
```D
struct Foo {
    int x;
}
```

.d:
```D
struct Foo {
    int x;
    // new in ....
    float y;
}
```

Reply:
Wow, I didn't know. IMO this makes not having changes to ABI somewhat understandable, but I'm still in favour of named arguments affecting the mangling.

After all, the example you just gave is likely to cause a lot of issues and it's likely that a DIP against it might be written at some point.

But you managed to convince me that there's no need to discuss that in the DIP.